File: ptolemy/codegen/c/actor/lib/Differential.c
/***CommonPreinitBlock($type)***/
$type $actorSymbol(_currentInput) = 0;
$type $actorSymbol(_lastInput) = 0;
/**/

/***CommonFireBlock***/
$actorSymbol(_currentInput) = $ref(input);
if ($actorSymbol(_lastInput) == 0) {
    $ref(output) = $actorSymbol(_currentInput);
} else {
    $ref(output) = $actorSymbol(_currentInput) - $actorSymbol(_lastInput);
}
$actorSymbol(_lastInput) = $actorSymbol(_currentInput);
/**/



/***TokenPreinitBlock***/
Token $actorSymbol(_currentInput);
Token $actorSymbol(_lastInput);
boolean $actorSymbol(isFirst) = true;
/**/

/***TokenFireBlock***/
$actorSymbol(_currentInput) = $ref(input);
if ($actorSymbol(isFirst)) {
    $ref(output) = $actorSymbol(_currentInput);
    $actorSymbol(isFirst) = false;
} else {
    $ref(output) = $tokenFunc($actorSymbol(_currentInput)::subtract($actorSymbol(_lastInput)));
}
$actorSymbol(_lastInput) = $actorSymbol(_currentInput);
/**/

File: ptolemy/actor/lib/Differential.java
/* An actor that outputs the difference between successive inputs.

 Copyright (c) 1998-2014 The Regents of the University of California.
 All rights reserved.
 Permission is hereby granted, without written agreement and without
 license or royalty fees, to use, copy, modify, and distribute this
 software and its documentation for any purpose, provided that the above
 copyright notice and the following two paragraphs appear in all copies
 of this software.

 IN NO EVENT SHALL THE UNIVERSITY OF CALIFORNIA BE LIABLE TO ANY PARTY
 FOR DIRECT, INDIRECT, SPECIAL, INCIDENTAL, OR CONSEQUENTIAL DAMAGES
 ARISING OUT OF THE USE OF THIS SOFTWARE AND ITS DOCUMENTATION, EVEN IF
 THE UNIVERSITY OF CALIFORNIA HAS BEEN ADVISED OF THE POSSIBILITY OF
 SUCH DAMAGE.

 THE UNIVERSITY OF CALIFORNIA SPECIFICALLY DISCLAIMS ANY WARRANTIES,
 INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
 MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE. THE SOFTWARE
 PROVIDED HEREUNDER IS ON AN "AS IS" BASIS, AND THE UNIVERSITY OF
 CALIFORNIA HAS NO OBLIGATION TO PROVIDE MAINTENANCE, SUPPORT, UPDATES,
 ENHANCEMENTS, OR MODIFICATIONS.

 PT_COPYRIGHT_VERSION_2
 COPYRIGHTENDKEY

 */
package ptolemy.actor.lib;

import ptolemy.data.Token;
import ptolemy.kernel.CompositeEntity;
import ptolemy.kernel.util.IllegalActionException;
import ptolemy.kernel.util.NameDuplicationException;

///////////////////////////////////////////////////////////////////
//// Differential

/**
 Output the current input minus the previous input, or if there
 has been no previous input, the current input itself.

 @author Edward A. Lee
 @version $Id$
 @since Ptolemy II 2.0
 @Pt.ProposedRating Red (eal)
 @Pt.AcceptedRating Red (cxh)
 */
public class Differential extends Transformer {
    /** Construct an actor with the given container and name.
     *  @param container The container.
     *  @param name The name of this actor.
     *  @exception IllegalActionException If the actor cannot be contained
     *   by the proposed container.
     *  @exception NameDuplicationException If the container already has an
     *   actor with this name.
     */
    public Differential(CompositeEntity container, String name)
            throws NameDuplicationException, IllegalActionException {
        super(container, name);
    }

    ///////////////////////////////////////////////////////////////////
    ////                         public methods                    ////

    /** Consume at most one token from the <i>input</i> port and output
     *  its value minus the value of the input read in the previous
     *  iteration.  If there has been no previous iteration, then
     *  output the current input.  If there is no input, then produce
     *  no output.
     *  @exception IllegalActionException If subtraction is not
     *   supported by the supplied tokens.
     */
    @Override
    public void fire() throws IllegalActionException {
        super.fire();
        if (input.hasToken(0)) {
            _currentInput = input.get(0);

            if (_lastInput != null) {
                output.broadcast(_currentInput.subtract(_lastInput));
            } else {
                output.broadcast(_currentInput);
            }
        }
    }

    /** Reset to indicate that no input has yet been seen.
     *  @exception IllegalActionException If the parent class throws it.
     */
    @Override
    public void initialize() throws IllegalActionException {
        super.initialize();
        _lastInput = null;
    }

    /** Record the most recent input as the latest input.
     *  @exception IllegalActionException If the base class throws it.
     */
    @Override
    public boolean postfire() throws IllegalActionException {
        _lastInput = _currentInput;
        return super.postfire();
    }

    ///////////////////////////////////////////////////////////////////
    ////                         private members                   ////
    private Token _currentInput;

    private Token _lastInput;
}
